pilopt/src/qse_opt.rs

```diff
@@ -0,0 +1,135 @@
+use std::iter;
+
+use itertools::Itertools;
+use powdr_ast::analyzed::{
+    AlgebraicBinaryOperation, AlgebraicBinaryOperator, AlgebraicExpression, AlgebraicReference,
+    AlgebraicUnaryOperation, AlgebraicUnaryOperator, Analyzed, Identity, PolynomialIdentity,
+};
+use powdr_constraint_solver::{
+    quadratic_symbolic_expression::QuadraticSymbolicExpression,
+    solver::{self, SolveResult},
+    symbolic_expression::{BinaryOperator, SymbolicExpression, UnaryOperator},
+};
+use powdr_number::FieldElement;
+
+pub fn run_qse_optimization<T: FieldElement>(pil_file: &mut Analyzed<T>) {
+    let identities = pil_file
+        .identities
+        .iter()
+        .filter_map(|identity| match identity {
+            Identity::Polynomial(PolynomialIdentity { expression, .. }) => {
+                try_algebraic_to_quadratic_symbolic_expression(expression)
+            }
+            _ => None,
+        })
+        .collect_vec();
+
+    match solver::Solver::new(identities).solve() {
+        Err(_) => {
+            log::error!("Error while QSE-optimizing. This is usually the case when the constraints are inconsistent.");
+        }
+        Ok(SolveResult {
+            simplified_algebraic_constraints,
+            ..
+        }) => {
+            let mut replacements = 0;
+            // TODO this is not correct if the are constraints with publics and challenges, because we skip them.
+            for identity in &mut pil_file.identities {
+                if let Identity::Polynomial(PolynomialIdentity { expression, .. }) = identity {
+                    if let Some(constraint) = simplified_algebraic_constraints.get(replacements) {
+                        *expression = quadratic_symbolic_expression_to_algebraic(constraint);
+                    } else {
+                        *expression = AlgebraicExpression::Number(0.into());
```

Comment:
I don't understand why this would happen? In the end, we should have `replacements == simplified_algebraic_constraints.len()`, no? The solver wouldn't remove any constraints?

Reply:
Why wouldn't it? If there are two equivalent constraints? If we keep the interface flexible (some constraints in, some constraints out), we won't have to change this part later on if we choose to add or remove constraints.

Reply:
Anyway, I agree with the implementation you have now :)